### PR TITLE
Round up amount before casting to int

### PR DESCRIPTION
**Pull request by dumega**
# Description
To make the casting more accurate, add `round` function before casting.

BTW. actually now we have `round` fiction for some casting while the rest not

Fixes: https://app.asana.com/0/941895179897714/1149342215992761/f

(Optional: Add a changelog by writing a comment after "#changelog" on the next line)

\#changelog Correct issue in discount rounding

# Type of change

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Asana task link and provided a changelog message if applicable.

### CODE REVIEW

File: app/code/community/Bolt/Boltpay/Model/Order.php [hunk 1]
```diff
@@ -683,7 +683,7 @@ protected function validateTotals(Mage_Sales_Model_Quote $immutableQuote, $trans
 
         if ($transaction->shouldDoTaxTotalValidation) {
             $magentoTaxTotal = (int)(( @$magentoTotals['tax']) ? round($magentoTotals['tax']->getValue() * 100) : 0 );
-            $boltTaxTotal = (int)$transaction->order->cart->tax_amount->amount;
+            $boltTaxTotal = (int)round($transaction->order->cart->tax_amount->amount);
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/Model/Order.php]: No rounding is needed or desired here.  This comes from Bolt directly.

File: app/code/community/Bolt/Boltpay/Model/BoltOrder.php [hunk 2]
```diff
@@ -473,8 +473,8 @@ protected function getTaxForAdmin( $taxTotal ) {
     protected function getCorrectedTotal($projectedTotal, $magentoDerivedCartData)
     {
         // we'll check if we can simply dividing by two corrects the problem
-        if ($projectedTotal == (int)($magentoDerivedCartData['total_amount']/2)) {
-            $magentoDerivedCartData["total_amount"] = (int)($magentoDerivedCartData['total_amount']/2);
+        if ($projectedTotal == (int)round($magentoDerivedCartData['total_amount']/2)) {
+            $magentoDerivedCartData["total_amount"] = (int)round($magentoDerivedCartData['total_amount']/2);
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/Model/BoltOrder.php]: There is no proof that this is needed.  It falls outside of the scope of what you identified as a problem.  Please limit your changes to validation step.

File: app/code/community/Bolt/Boltpay/Model/Order.php [hunk 3]
```diff
@@ -659,8 +659,8 @@ protected function validateTotals(Mage_Sales_Model_Quote $immutableQuote, $trans
 
             if ( $cartItem->shouldNotBeValidated ){ continue; }
 
-            $boltPrice = (int)$boltCartItem->total_amount->amount;
-            $magentoRowPrice = (int) ( $cartItem->getRowTotalWithDiscount() * 100 );
+            $boltPrice = (int)round($boltCartItem->total_amount->amount);
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/Model/Order.php]: No rounding is needed or desired here. This comes from Bolt directly.

File: app/code/community/Bolt/Boltpay/Model/Order.php [hunk 4]
```diff
@@ -659,8 +659,8 @@ protected function validateTotals(Mage_Sales_Model_Quote $immutableQuote, $trans
...
+            $magentoRowPrice = (int)round( $cartItem->getRowTotalWithDiscount() * 100 );
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/Model/Order.php]: Is this necessary?  Does `$cartItem->getRowTotalWithDiscount()` return a value beyond two significant decimal digits?

File: app/code/community/Bolt/Boltpay/Model/Order.php [hunk 5]
```diff
@@ -701,8 +701,8 @@ protected function validateTotals(Mage_Sales_Model_Quote $immutableQuote, $trans
         if ($transaction->shouldSkipDiscountAndShippingTotalValidation) return;
 
         if ($transaction->shouldDoDiscountTotalValidation) {
-            $magentoDiscountTotal = (int)(($immutableQuote->getBaseSubtotal() - $immutableQuote->getBaseSubtotalWithDiscount()) * 100);
-            $boltDiscountTotal = (int)$transaction->order->cart->discount_amount->amount;
+            $magentoDiscountTotal = (int)round(($immutableQuote->getBaseSubtotal() - $immutableQuote->getBaseSubtotalWithDiscount()) * 100);
+            $boltDiscountTotal = (int)round($transaction->order->cart->discount_amount->amount);
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/Model/Order.php]: No rounding is needed or desired here. This comes from Bolt directly.

File: app/code/community/Bolt/Boltpay/Model/Order.php [hunk 6]
```diff
@@ -719,8 +719,8 @@ protected function validateTotals(Mage_Sales_Model_Quote $immutableQuote, $trans
 
         if ( $transaction->shouldDoShippingTotalValidation && !$immutableQuote->isVirtual() ) {
             $shippingAddress = $immutableQuote->getShippingAddress();
-            $magentoShippingTotal = (int) (($shippingAddress->getShippingAmount() - $shippingAddress->getBaseShippingDiscountAmount()) * 100);
-            $boltShippingTotal = (int)$transaction->order->cart->shipping_amount->amount;
+            $magentoShippingTotal = (int)round(($shippingAddress->getShippingAmount() - $shippingAddress->getBaseShippingDiscountAmount()) * 100);
+            $boltShippingTotal = (int)round($transaction->order->cart->shipping_amount->amount);
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/Model/Order.php]: No rounding is needed or desired here. This comes from Bolt directly.

File: app/code/community/Bolt/Boltpay/controllers/ApiController.php [hunk 7]
```diff
@@ -262,7 +262,7 @@ public function create_orderAction() {
                 array(
                     'status' => 'success',
                     'display_id' => $order->getIncrementId(),
-                    'total' => (int)($order->getGrandTotal() * 100),
+                    'total' => (int)round($order->getGrandTotal() * 100),
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/controllers/ApiController.php]: Is this necessary? Does `$order->getGrandTotal()` return a value beyond two significant decimal digits?  Can we get an output showing a number here like "4.5943"?   I think that it will always be "4.59".

File: app/code/community/Bolt/Boltpay/controllers/ConfigurationController.php [hunk 8]
```diff
@@ -136,7 +136,7 @@ protected function checkPublishableKey($key)
         try{
             $client = $this->boltHelper()->getApiClient();
             $response = $client->get($url,$headerInfo)->getStatusCode();
-            return (int)($response / 100) == 2;
+            return (int)round($response / 100) == 2;
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/controllers/ConfigurationController.php]: This is unrelated, unnecessary, and wrong.  This is not related to discount values and is checking HTTP response codes.  Please revert this back to the original form.

**Comment by dumega** [comment]
@mccottry  I have reverted unrelated changes
